gitpod-io/gitpod
# [installer]: create tests to verify render result

## Description

This is a proposed change to allow for testing of the rendered YAML in the Installer.

The data is stored in `cmd/testdata/render`, with the test names being stored in `cmd/testdata/render/<name>/<config|output>.golden`. Files ending `.golden` are excluded from the hooks to make them testable. The intention is that there will be a suite of these tests and they verify that the known output is matched. When a change is made (even outside of this directory), the files will require changing and @gitpod-io/engineering-self-hosted is notified about potentially breaking changes to the rendered YAML - we can then approve (or not) the changes.

How to generate the files is documented.

Randomisation is controlled by specifying the `Seed` for `math/rand` and by giving the `crypto/rand` a known string. A warning is printed to stderr if this is enabled during the YAML rendering process, which is designed to act as a deterrent to anyone using this in production.

### Risks Introduced

It's not impossible for a PR to pass its tests in a branch and then fail when it's merged. This can be mitigated by getting into the habit of doing a final `git rebase main` before this is merged.

In the event of a failure in the `main` branch, the engineer who authored the original PR is responsible for ensuring that this is fixed in a timely fashion.

## Related Issue(s)

Fixes #11147

## How to test

Run tests in Werft. See also the [docs](https://github.com/gitpod-io/gitpod/blob/sje/render-testing/install/installer/cmd/testdata/render/README.md)

## Release Notes

```release-note
[installer]: create tests to verify render result
```

## Documentation



## Werft options:
